cpp/support/thread_safe_cache.h

```diff
@@ -179,6 +188,228 @@ class ThreadSafeCache<Key, Value> {
   std::shared_mutex erase_mutex_;
 };
 
+namespace details {
+
+template <typename Value, typename... Keys>
+class LRUCacheImpl {
+ protected:
+  struct LRUnode;
+  using KeySet = std::variant<const Keys*...>;
+  using LRUlist = std::list<LRUnode>;
```

Comment:
Don't use std::list. Implement a linked list with **minimal functionality** using std::vector<T> pool and std::vector<int> free_nodes_ in support/container.h. You can also check out  StackElementBuffer

See also 
https://baptiste-wicht.com/posts/2012/12/cpp-benchmark-vector-list-deque.html